force-app/main/service/DonationHistoryService.cls

```diff
@@ -0,0 +1,191 @@
+        donationHistory.addAll(partialAndSoftCredits);
+
+        return donationHistory;
+    }
+    
+    /** 
+    * @author Salesforce.org
+    * @date 2021
+    * @description Verify hard credit condition.
+    */
+    private static List<DonationHistoryController.DonationHistory> buildHardCreditHistoryFromOCRs(List<OpportunityContactRole> credits){
+        List<DonationHistoryController.DonationHistory> hardCreditList = new List<DonationHistoryController.DonationHistory>();
+        for(OpportunityContactRole ocr: credits){
+            if (ocr.isPrimary) {
+                DonationHistoryController.DonationHistory hardCredit = new DonationHistoryController.DonationHistory(ocr, HARD_CREDIT);
+                hardCreditList.add(hardCredit);
+            }
+        }
+        return hardCreditList;
+    }
+    
+    /** 
+    * @author Salesforce.org
+    * @date 2021
+    * @description Verify soft credit condition.
+    */
+    private static List<DonationHistoryController.DonationHistory> buildSoftCreditHistoryFromOCRs(List<OpportunityContactRole> donations){
+        //with the list of donations verify if the Role match the filter rule value on the NPSP_ClosedWon_Opps_SC filter group 
+        Filter_Rule__mdt roles = CMT_FilterRule_SEL.getFilterRuleByDeveloperName(NPSP_CREDIT_ROLES);
+        List<String> softCreditRoles = roles.Constant__c.split(';');
+        List<DonationHistoryController.DonationHistory> softCreditList = new List<DonationHistoryController.DonationHistory>();
+        for(OpportunityContactRole ocr: donations){
+            if (softCreditRoles.contains(ocr.Role) && ocr.isPrimary==false) {
+                DonationHistoryController.DonationHistory softCredit = new DonationHistoryController.DonationHistory(ocr, SOFT_CREDIT);
+                softCreditList.add(softCredit);
+            }
+        }
+        return softCreditList;
+    }
+
+    /** 
+    * @author Salesforce.org
+    * @date 2021
+    * @description Filter for ids already classified as Soft Credit.
+    */
+    private static List<Id> filterVerifiedIds(List<DonationHistoryController.DonationHistory> softCredits){
+        List<Id> softCreditIds = new List<Id>();
+        for(DonationHistoryController.DonationHistory dh: softCredits){
+            softCreditIds.add(dh.id);
+        }
+        return softCreditIds;
+    }
+
+    /** 
+    * @author Salesforce.org
+    * @date 2021
+    * @description Verify partial soft credit condition.
+    */
+    private static List<DonationHistoryController.DonationHistory> verifyPartialSoftCreditCondition(List<Id> softCreditIds, List<OpportunityContactRole> credits){
+        List<Id> partialSoftCreditIds = new List<Id>();
+        List<Partial_Soft_Credit__c> partialSoftCredit = new DonationHistorySelector().getPartialSoftCredit(softCreditIds);
+        for(Partial_Soft_Credit__c ps: partialSoftCredit){
+            partialSoftCreditIds.add(ps.Contact_Role_ID__c);
+        }
```

Comment:
I think this is where we can apply an instance of the CMT_FilterRule class.

```suggestion
        Filter_Rule__mdt ruleCmt = CMT_FilterRule_SEL.getFilterRuleByDeveloperName(NPSP_CREDIT_ROLES);
        CMT_FilterRule rule = new CMT_FilterRule(ruleCmt);
        for(Partial_Soft_Credit__c ps: partialSoftCredit) {
            if(rule.isIncluded(ps)) {
                partialSoftCreditIds.add(ps.Contact_Role_ID__c);
            }
        }
```

Then, when the partialSoftCreditIds variable is referenced further down, it will only contain Partial Soft Credits with a valid Role. The Role on the OCR and on the Role Name field on Partial Soft Credit should be the same value. With this change, we should not need to do any filtering on the Role/Role Name field in `buildSoftCreditHistoryFromOCRs`

Reply:
instance of CMT_FilterRule added!